File: 02. Two Pointers and Sliding Window/03. Longest Subtring without repeating characters/main.js
function bruteForce(s) {
    let maxLength = 0;
    for (let i = 0; i < s.length; i++) {
        for (let j = i; j < s.length; j++) {
            if (allUnique(s, i, j)) {
                maxLength = Math.max(maxLength, j - i + 1);
            }
        }
    }
    return maxLength;
}

function allUnique(s, start, end) {
    let set = new Set();
    for (let i = start; i <= end; i++) {
        if (set.has(s[i])) return false;
        set.add(s[i]);
    }
    return true;
}

function improvedBruteForce(s) {
    let maxLength = 0;
    for (let i = 0; i < s.length; i++) {
        let seen = new Set();
        for (let j = i; j < s.length; j++) {
            if (seen.has(s[j])) break;
            seen.add(s[j]);
            maxLength = Math.max(maxLength, j - i + 1);
        }
    }
    return maxLength;
}

function slidingWindow(s) {
    let lastSeen = new Map(), left = 0, maxLength = 0;
    for (let right = 0; right < s.length; right++) {
        if (lastSeen.has(s[right])) {
            left = Math.max(left, lastSeen.get(s[right]) + 1);
        }
        lastSeen.set(s[right], right);
        maxLength = Math.max(maxLength, right - left + 1);
    }
    return maxLength;
}

function slidingWindowSet(s) {
    let seen = new Set(), left = 0, maxLength = 0;
    for (let right = 0; right < s.length; right++) {
        while (seen.has(s[right])) {
            seen.delete(s[left]);
            left++;
        }
        seen.add(s[right]);
        maxLength = Math.max(maxLength, right - left + 1);
    }
    return maxLength;
}

function main() {
    const s = prompt("Enter a string:");
    console.log("Brute Force:", bruteForce(s));
    console.log("Improved Brute Force:", improvedBruteForce(s));
    console.log("Sliding Window (Hash Map):", slidingWindow(s));
    console.log("Sliding Window (Hash Set):", slidingWindowSet(s));
}

main();

File: 02. Two Pointers and Sliding Window/03. Longest Subtring without repeating characters/main.java
import java.util.*;

public class main {
    
    // 1. Brute Force (O(n^3))
    public static int bruteForce(String s) {
        int n = s.length(), maxLength = 0;
        for (int i = 0; i < n; i++) {
            for (int j = i; j < n; j++) {
                if (allUnique(s, i, j)) {
                    maxLength = Math.max(maxLength, j - i + 1);
                }
            }
        }
        return maxLength;
    }

    private static boolean allUnique(String s, int start, int end) {
        Set<Character> set = new HashSet<>();
        for (int i = start; i <= end; i++) {
            if (set.contains(s.charAt(i))) return false;
            set.add(s.charAt(i));
        }
        return true;
    }

    // 2. Improved Brute Force (O(n^2))
    public static int improvedBruteForce(String s) {
        int maxLength = 0;
        for (int i = 0; i < s.length(); i++) {
            Set<Character> seen = new HashSet<>();
            for (int j = i; j < s.length(); j++) {
                if (seen.contains(s.charAt(j))) break;
                seen.add(s.charAt(j));
                maxLength = Math.max(maxLength, j - i + 1);
            }
        }
        return maxLength;
    }

    // 3. Sliding Window (O(n)) using HashMap
    public static int slidingWindow(String s) {
        Map<Character, Integer> lastSeen = new HashMap<>();
        int left = 0, maxLength = 0;
        
        for (int right = 0; right < s.length(); right++) {
            if (lastSeen.containsKey(s.charAt(right))) {
                left = Math.max(left, lastSeen.get(s.charAt(right)) + 1);
            }
            lastSeen.put(s.charAt(right), right);
            maxLength = Math.max(maxLength, right - left + 1);
        }
        
        return maxLength;
    }

    // 4. Sliding Window using HashSet
    public static int slidingWindowSet(String s) {
        Set<Character> seen = new HashSet<>();
        int left = 0, maxLength = 0;
        
        for (int right = 0; right < s.length(); right++) {
            while (seen.contains(s.charAt(right))) {
                seen.remove(s.charAt(left));
                left++;
            }
            seen.add(s.charAt(right));
            maxLength = Math.max(maxLength, right - left + 1);
        }
        
        return maxLength;
    }

    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        System.out.print("Enter a string: ");
        String s = scanner.next();
        
        System.out.println("Brute Force: " + bruteForce(s));
        System.out.println("Improved Brute Force: " + improvedBruteForce(s));
        System.out.println("Sliding Window (Hash Map): " + slidingWindow(s));
        System.out.println("Sliding Window (Hash Set): " + slidingWindowSet(s));

        scanner.close();
    }
}
